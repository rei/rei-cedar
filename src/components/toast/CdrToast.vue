<script>
import {
  useCssModule, defineComponent, computed, ref, watch, onUpdated,
} from 'vue';
import propValidator from '../../utils/propValidator';
import IconXSm from '../icon/comps/x-sm.vue';
import CdrButton from '../button/CdrButton.vue';

export default defineComponent({
  name: 'CdrToast',
  components: {
    IconXSm,
    CdrButton,
  },
  props: {
    type: {
      type: String,
      validator: (value) => propValidator(
        value,
        ['info', 'warning', 'success', 'error', 'default'],
      ),
      default: 'default',
    },
    open: {
      type: Boolean,
      default: false,
    },
    autoDismiss: {
      type: Boolean,
      default: true,
    },
    dismissDelay: {
      type: Number,
      default: 5000,
    },
  },
  emits: ['open', 'closed'],
  setup(props, ctx) {
    const baseClass = 'cdr-toast';
    const style = useCssModule();
    const hasIconLeft = ctx.slots['icon-left'];
    const opened = ref(null);
    const toastEl = ref(null);
    let timeout;
    let toastElement;

    const typeClass = computed(() => props.type && `${baseClass}--${props.type}`);

    const openToast = (e) => {
      if (timeout) {
        clearTimeout(timeout);
      } else {
        ctx.emit('open', e);
      }
      opened.value = true;
      if (props.autoDismiss && !e) {
        closeToastWithDelay();
      }
    };

    const closeToast = (e) => {
      removeHandlers();
      opened.value = false;
      ctx.emit('closed', e);
    };

    const closeToastWithDelay = (e) => {
      timeout = setTimeout(() => {
        removeHandlers();
        opened.value = false;
        ctx.emit('closed', e);
      }, props.dismissDelay);
    };

    const addHandlers = () => {
      toastElement = toastEl.value;
      if (toastElement) {
        toastElement.addEventListener('mouseover', openToast);
        toastElement.addEventListener('mouseleave', closeToastWithDelay);
      }
    };

    const removeHandlers = () => {
      if (toastElement) {
        toastElement.removeEventListener('mouseover', openToast);
        toastElement.removeEventListener('mouseleave', closeToastWithDelay);
      }
    };

    watch(() => props.open, () => {
      if (props.open) openToast();
    });

    onUpdated(() => {
      if (props.autoDismiss) addHandlers();
    });

    return {
      baseClass,
      style,
      hasIconLeft,
      opened,
      toastEl,
      timeout,
      toastElement,
      typeClass,
      openToast,
      closeToast,
      closeToastWithDelay,
      addHandlers,
      removeHandlers,
    };
  },
});
</script>

<template>
  <transition
    :enter-from-class="style['cdr-toast__transition--toast-enter-from']"
    :enter-active-class="style['cdr-toast__transition--toast-enter-active']"
    :leave-to-class="style['cdr-toast__transition--toast-leave-to']"
    :leave-active-class="style['cdr-toast__transition--toast-leave-active']"
  >
    <div
      v-if="opened"
      :class="[style[baseClass], style[typeClass]]"
      role="status"
      ref="toastEl"
    >
      <div :class="[style['cdr-toast__main']]">
        <div
          v-if="hasIconLeft"
          :class="[style['cdr-toast__icon-left']]"
        >
          <slot name="icon-left" />
        </div>
        <span :class="[style['cdr-toast__message']]">
          <slot name="default" />
        </span>
        <cdr-button
          :class="[style['cdr-toast__close-button']]"
          icon-only
          @click="closeToast"
          aria-label="Close"
          size="small"
        >
          <slot name="icon">
            <icon-x-sm
              inherit-color
            />
          </slot>

        </cdr-button>
      </div>
    </div>
  </transition>
</template>

<style lang="scss" module src="./styles/CdrToast.module.scss">
</style>
