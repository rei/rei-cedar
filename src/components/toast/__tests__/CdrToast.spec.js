import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrToast from '../CdrToast.vue';

describe('CdrToast', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(CdrToast, {
      slots: {
        default: 'toast content',
        'icon-left': '<div />',
      },
    });
  });
  it('matches snapshot', async () => {
    wrapper.setProps({ open: true });
    //wrapper.setData({ opened: true });
    await wrapper.vm.$nextTick(() => {
      expect(wrapper.element).toMatchSnapshot();
    })
  });
  it('handles close triggers', async () => {
    const spyCloseToast = vi.spyOn(wrapper.vm, 'closeToast');
    wrapper.setProps({ open: true });
    await wrapper.vm.$nextTick(() => {
      wrapper.find('button').trigger('click');
      expect(spyCloseToast).toBeCalled();
    });
  });
  //To Do: This one is still not working correctly
  // it('closes after 5 seconds', async () => {
  //   vi.useFakeTimers();
  //   const spyCloseToastWithDelay = vi.spyOn(wrapper.vm, 'closeToastWithDelay');
  //   wrapper.setProps({ open: true });
  //   vi.advanceTimersByTime(5000),
  //   await wrapper.vm.$nextTick(() => {
  //     expect(spyCloseToastWithDelay).toBeCalled();
  //   });
  // });
});