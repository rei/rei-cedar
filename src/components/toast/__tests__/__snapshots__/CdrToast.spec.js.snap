// Vitest Snapshot v1

exports[`CdrToast > matches snapshot 1`] = `
<transition-stub>
  <div
    class="cdr-toast cdr-toast--default"
    role="status"
  >
    <div
      class="cdr-toast__main"
    >
      <div
        class="cdr-toast__icon-left"
      >
        
        <div />
        
      </div>
      <span
        class="cdr-toast__message"
      >
        
        toast content
        
      </span>
      <button
        aria-label="Close"
        class="cdr-button cdr-button--primary cdr-button--icon-only-small cdr-button--icon-only cdr-button--icon-only cdr-toast__close-button"
        type="button"
      >
        
        
        
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-icon--medium cdr-icon--inherit-color"
          slot="icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          
          
          <path
            d="M13.406 12.006l3.297-3.296a1 1 0 10-1.414-1.414l-3.297 3.295-3.285-3.295A1 1 0 107.293 8.71l3.285 3.295-3.285 3.288a1 1 0 001.414 1.415l3.285-3.289 3.297 3.289a1 1 0 001.414-1.415l-3.297-3.287z"
            role="presentation"
          />
          
          <!--v-if-->
        </svg>
        
        
        
      </button>
    </div>
  </div>
</transition-stub>
`;

exports[`CdrToast matches snapshot 1`] = `
<transition-stub>
  <div
    class="cdr-toast cdr-toast--default"
    role="status"
  >
    <div
      class="cdr-toast__main"
    >
      <div
        class="cdr-toast__icon-left"
      >
        
        <div />
        
      </div>
      <span
        class="cdr-toast__message"
      >
        
        toast content
        
      </span>
      <button
        aria-label="Close"
        class="cdr-button cdr-button--primary cdr-button--icon-only-small cdr-button--icon-only cdr-button--icon-only cdr-toast__close-button"
        type="button"
      >
        
        
        
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-icon--medium cdr-icon--inherit-color"
          slot="icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          
          
          <path
            d="M13.406 12.006l3.297-3.296a1 1 0 10-1.414-1.414l-3.297 3.295-3.285-3.295A1 1 0 107.293 8.71l3.285 3.295-3.285 3.288a1 1 0 001.414 1.415l3.285-3.289 3.297 3.289a1 1 0 001.414-1.415l-3.297-3.287z"
            role="presentation"
          />
          
          <!--v-if-->
        </svg>
        
        
        
      </button>
    </div>
  </div>
</transition-stub>
`;
